### ISSUE TITLE
instanceof Promise in ternary does not narrow Promiselike in 4.1

### ISSUE THREAD

**Issue by brainkim** [microsoft/TypeScript#41606]
**TypeScript Version:**  4.1.0-beta

**Search Terms:**
promise promiselike instanceof
**Code**

```ts repro
declare const p: PromiseLike<number>;
const p1 = p instanceof Promise ? p : Promise.resolve(p);
const p2: Promise<number> = p1.then((v) => v + 1);
```

**Expected behavior:**

No errors.

**Actual behavior:**
p2 errors because the expression has been determined to be promise-like.
Also p1 is inferred as PromiseLike<number> & Promise<any>.

**Playground Link:** 
https://www.typescriptlang.org/play?ts=4.1.0-beta#code/CYUwxgNghgTiAEYD2A7AzgF3gBwFzwAUYkBbASzRABkyBrEAHhQFcSAjEGAPgG4AoZOizYAjPAC8OeGSFQUYEEgBmhYuUrwA-FPxFSFEADo4aJBABuIABTYAlP0GYcAJl1qDTVh24ScIwxgAFiAoVlbmthJc8ObwANTwIvZAA

**Related Issues:**

**Comment by MartinJohns** [comment]
> const p2: Promise<number> = p.then((v) => v + 1);

Did you mean to use `p1` instead of `p`?

**Comment by brainkim** [comment]
Ah yeah. I was wondering why the bug is only on the new version and I couldn’t reproduce the old behavior. Fixed the typescript example which only fails on 4.1. Edited original.

**Comment by andrewbranch** [comment]
I think this cooould be an instance of #41563. It’s essentially an overload selection problem—the intersected `PromiseLike` and `Promise` is a reasonably correct representation of what we know about `p`, but the problem is that we now have an intersected signature for `then`, which equates to overloads, and no heuristics to help us pick between them.

**Comment by brainkim** [comment]
@rubiesonthesky This appears to have been fixed since 4.2.2. Thanks.
